stdlib/public/SDK/Foundation/CheckClass.mm

```diff
@@ -0,0 +1,57 @@
+#import <Foundation/Foundation.h>
+#import <CoreGraphics/CoreGraphics.h>
+
+#include <swift/Runtime/Metadata.h>
+#include <stdio.h>
+
+
+@interface NSKeyedUnarchiver (SwiftAdditions)
++ (int)_swift_checkClassAndWarnForKeyedArchiving:(Class)cls
+     NS_SWIFT_NAME(_swift_checkClassAndWarnForKeyedArchiving(_:));
+@end
+
+@implementation NSKeyedUnarchiver (SwiftAdditions)
+
+/// Checks if class \p cls is good for archiving.
+///
+/// If not, a runtime warning is printed.
+///
+/// \return Returns the status
+/// 0:       not a problem class (either non-Swift or has an explicit name)
+/// 1:       a Swift generic class
+/// 2:       a Swift non-generic class where adding @objc is valid
+/// default: a Swift class with some other problem that means it shouldn't be
+///          archived
++ (int)_swift_checkClassAndWarnForKeyedArchiving:(Class)cls {
+  const swift::ClassMetadata *theClass = (swift::ClassMetadata *)cls;
+
+  // Is it a (real) swift class?
+  if (!theClass->isTypeMetadata() || theClass->isArtificialSubclass())
```

Comment:
On future OS versions this code will believe that OS-Swift classes are "not Swift".

One way for this code to recognize future OS versions would be for it to look for another method on NSKeyedUnarchiver that does not yet exist. The future Foundation would implement that method, and this code would call it if it exists.